### PR TITLE
Reduce the load of "gfollower"

### PR DESCRIPTION
**Pull request by annando**
The "gfollower" functionality that had been introduced lately (that is planned to replace glink) had shown to be a performance problem for small sites. (For example currently my system has got more than 800,000 open workerqueue jobs for gcontact discovery)

To reduce the problem the action to activate that option is now a select box. You can select between no discovery, only the discovery of your contacts and the full blown recursive discovery that will eventually lead to a database with all users in the world.

Since even the second option does generate some load (currently around 180,000 open jobs on my other server), another functionality is introduced that doesn't need any additional background processing: We simply collecting who is interacting with whom. This is partially (and in an ugly way) be done with the `glink` table as well. But this here is more straightforward and with no overhead.

In the future (hopefully in the next version) we can then have a look at the data and can help the manual and automatic discovery of new contacts with the one or the other data source. (and then we can remove `glink` (and possibly some more stuff from "PoCo").

### CODE REVIEW

**Comment by AlfredSK** [comment]
Wow. 800000. Is it this option you're referring to: 'Discover followers/followings from global contacts'?

I didn't know it exists. 😜 Looks like it's disabled by default.

**Comment by annando** [comment]
@AlfredSK yeah, exactly that option. I coded it last week, short before the RC. Now, after several days I realized that it only is a viable option on real huge servers with a high number of parallel workers. But since we need some replacement for the `glink` table, the other table took place. It's not as accurate as the other table that shows the true followers, but it should be accurate enough to later create some follower suggestion out of it.

File: src/Model/Item.php [hunk 1]
```diff
@@ -1750,6 +1750,11 @@ public static function insert($item, $force_parent = false, $notify = false, $do
 					DBA::update('thread', ['mention' => true], ['iid' => $parent_id]);
 					Logger::log('tagged thread ' . $parent_id . ' as mention for user ' . $item['uid'], Logger::DEBUG);
 				}
+
+				// Update the contact relations
+				if ($item['author-id'] != $parent['author-id']) {
+					dba::update('contact-relation', ['last-interaction' => $item['created']], ['cid' => $parent['author-id'], 'relation-cid' => $item['author-id']], true);
```
Comment (MrPetovan) [inline src/Model/Item.php]:
```suggestion
					DBA::update('contact-relation', ['last-interaction' => $item['created']], ['cid' => $parent['author-id'], 'relation-cid' => $item['author-id']], true);
```

File: src/Module/Admin/Site.php [hunk 2]
```diff
@@ -178,7 +179,7 @@ function update_table(App $a, $table_name, $fields, $old_url, $new_url)
 		$optimize_max_tablesize = (!empty($_POST['optimize_max_tablesize']) ? intval(trim($_POST['optimize_max_tablesize'])) : 100);
 		$optimize_fragmentation = (!empty($_POST['optimize_fragmentation']) ? intval(trim($_POST['optimize_fragmentation'])) : 30);
 		$poco_completion        = (!empty($_POST['poco_completion'])        ? intval(trim($_POST['poco_completion']))        : false);
-		$gcontact_discovery     = (!empty($_POST['gcontact_discovery'])     ? intval(trim($_POST['gcontact_discovery']))     : false);
+		$gcontact_discovery     = (!empty($_POST['gcontact_discovery'])     ? intval(trim($_POST['gcontact_discovery']))     : 0);
```
Comment (MrPetovan) [inline src/Module/Admin/Site.php]: Please use the `GContact::DISCOVERY_NONE` constant you introduced.

File: src/Module/Admin/Site.php [hunk 3]
```diff
@@ -671,7 +678,7 @@ public static function content(array $parameters = [])
...
 
 			'$poco_completion'        => ['poco_completion', DI::l10n()->t('Periodical check of global contacts'), DI::config()->get('system', 'poco_completion'), DI::l10n()->t('If enabled, the global contacts are checked periodically for missing or outdated data and the vitality of the contacts and servers.')],
-			'$gcontact_discovery'     => ['gcontact_discovery', DI::l10n()->t('Discover followers/followings from global contacts'), DI::config()->get('system', 'gcontact_discovery'), DI::l10n()->t('If enabled, the global contacts are checked for new contacts among their followers and following contacts. This option will create huge masses of jobs, so it should only be activated on powerful machines.')],
+			'$gcontact_discovery'     => ['gcontact_discovery', DI::l10n()->t('Discover followers/followings from global contacts'), (string)intval(DI::config()->get('system', 'gcontact_discovery')), DI::l10n()->t('If enabled, the global contacts are checked for new contacts among their followers and following contacts. This option will create huge masses of jobs, so it should only be activated on powerful machines.'), $discovery_choices],
```
Comment (MrPetovan) [inline src/Module/Admin/Site.php]: Wait why do you have to do `(string)intval(...)` ?

Comment (annando) [reply src/Module/Admin/Site.php]: Copy&Paste :-)

Comment (MrPetovan) [reply src/Module/Admin/Site.php]: What other fields?

**Comment by AlfredSK** [comment]
Question: Is is normal that I get this message when I check if the database update and postupdate was successful...

```
sudo -u www-data php bin/console.php dbstructure update

sudo -u www-data php bin/console.php postupdate        
Check for pending update actions.
ALTER IGNORE TABLE `contact-relation` MODIFY `cid` int unsigned NOT NULL DEFAULT 0 COMMENT 'contact the related contact had interacted with', MODIFY `relation-cid` int unsigned NOT NULL DEFAULT 0 COMMENT 'related contact who had interacted with the contact'; 
Done.
Execute pending post updates.

All pending post updates are done.
```

The postupdate message doesn't go away. **I get the same message when I force-update the dbstructure.**

**Comment by annando** [comment]
Please try a `php bin/console.php dbstructure update -f`

**Comment by AlfredSK** [comment]
Ähem...

> **I get the same message when I force-update the dbstructure.**

Wich means literally the same message...

```
sudo -u www-data php bin/console.php dbstructure update -f
ALTER IGNORE TABLE `contact-relation` MODIFY `cid` int unsigned NOT NULL DEFAULT 0 COMMENT 'contact the related contact had interacted with', MODIFY `relation-cid` int unsigned NOT NULL DEFAULT 0 COMMENT 'related contact who had interacted with the contact'; 

```

**Comment by AlfredSK** [comment]
When I run the query on the database I get this:

```
Query OK, 0 rows affected (0.000 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

**Comment by annando** [comment]
Yeah, sorry. This was my typical writing comments while watching videos thing :-)

The fix is on it's way.

**Pull request merged**